src/services/cssFolding.ts

```diff
@@ -0,0 +1,83 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import { TextDocument, Position } from 'vscode-languageserver-types';
+import { TokenType, Scanner, IToken } from '../parser/cssScanner';
+import * as nodes from '../parser/cssNodes';
+import { ParseError, CSSIssueType } from '../parser/cssErrors';
+import * as languageFacts from './languageFacts';
+import { FoldingRangeList, FoldingRange, FoldingRangeType } from 'vscode-languageserver-protocol-foldingprovider';
+
+export function getFoldingRegions(document: TextDocument): FoldingRangeList {
+	function getStartLine(t: IToken) {
+		return document.positionAt(t.offset).line;
+	}
+	function getEndLine(t: IToken) {
+		return document.positionAt(t.offset + t.len).line;
+	}
+	function tokenToRange(t: IToken, type: FoldingRangeType = FoldingRangeType.Region): FoldingRange | null {
+		const startLine = getStartLine(t);
+		const endLine = getEndLine(t);
+
+		if (startLine !== endLine) {
+			return {
+				startLine,
+				endLine,
+				type
+			};
+		} else {
+			return null;
+		}
+	}
+
+	const ranges: FoldingRange[] = [];
+	const stack: number[] = [];
+
+	const scanner = new Scanner();
+	scanner.ignoreComment = false;
+	scanner.setSource(document.getText());
+
+	let token = scanner.scan();
+	while (token.type !== TokenType.EOF) {
+		const tt = TokenType[token.type];
+		switch(token.type) {
+			case TokenType.CurlyL: {
+				stack.push(getStartLine(token));
+				break;
+			}
+			case TokenType.CurlyR: {
+				if (stack.length !== 0) {
+					const startLine = stack.pop();
+					const endLine = getEndLine(token);
+					if (startLine !== endLine) {
```

Comment:
The current folding stye is to keep showing the opening '{' and the '}' and only fold what's in between.
-> endline = getEndLine(token) - 1

Reply:
I cannot do this because of such test cases:

```css
.foo {
  color: red }
```

The current way it works seems fine. I saw you added column to the folding protocol. We can make use of those in the next version of css folding.

Reply:
In that case the full last line stays.
It's a corner case...